### PR TITLE
feat(highlight): highlight array of terms instead of just one term

### LINKED ISSUE #2698: feature: NgbHighlight multiple terms
I'd like to use NgbHighlight to highlight multiple separate terms in the typeahead results. E.g. if searching for "foo bar" I'd like to be able to generate highlighted results that appear like: "**foo**sball **bar**ricade"

There doesn't seem to be any way to do that but I'd be happy to submit a PR to do so. (My plan is to make the 'term' input accept either a string or array of strings.) Is there anything I'm missing or any reason not to do that?

### PR DESCRIPTION
**Pull request by IAfanasov**
Closes #2698

Before submitting a pull request, please make sure you have at least performed the following:

 - [x] read and followed the [CONTRIBUTING.md](https://github.com/ng-bootstrap/ng-bootstrap/blob/master/CONTRIBUTING.md) guide.
 - [x] built and tested the changes locally.
 - [x] added/updated any applicable tests.
 - [x] added/updated any applicable API documentation.
 - [ ] added/updated any applicable demos.

### CODE REVIEW

**Comment by IAfanasov** [comment]
and this morning my mind came up with the idea of how to break this code
```
  fit('should highlight when term contains array with several items and one term is substring of another', () => {
    const fixture = createTestComponent(`<ngb-highlight result="foobar baz" [term]="['foo', 'foobar']"></ngb-highlight>`);

    expect(highlightHtml(fixture))
        .toBe('<span class="ngb-highlight">foobar</span> baz');
  });

  fit('should highlight when term contains array with several items and one term is substring of another', () => {
    const fixture = createTestComponent(`<ngb-highlight result="foobar baz" [term]="['foobar', 'foo']"></ngb-highlight>`);

    expect(highlightHtml(fixture))
        .toBe('<span class="ngb-highlight">foobar</span> baz');
  });
```

hope to find time to fix it this week

**Comment by IAfanasov** [comment]
Actually, I'm really not sure if it a bug. With several terms to highlight there could be some complex scenario and I feel like we should not resolve it automagically and better to rely on the order of terms provided by a developer.

Open to any ideas about it (:

File: src/typeahead/highlight.ts [hunk 1]
```diff
@@ -30,28 +30,30 @@ export class NgbHighlight implements OnChanges {
    * The text highlighting is added to.
    *
    * If the `term` is found inside this text, it will be highlighted.
+   * If the `term` contains array then all the items from it will be highlighted inside the text.
    */
-  @Input() result: string;
+  @Input() result: any;
```
Comment (maxokorokov) [inline src/typeahead/highlight.ts]: Don't see the necessity of changing this at the moment. 

The issue is about supporting multiple terms, also this is not covered by tests

File: src/typeahead/highlight.ts [hunk 2]
```diff
@@ -30,28 +30,30 @@ export class NgbHighlight implements OnChanges {
...
 
   /**
-   * The term to be highlighted.
+   * The term or array of terms to be highlighted.
    */
-  @Input() term: string;
+  @Input() term: any;
```
Comment (maxokorokov) [inline src/typeahead/highlight.ts]: Please use the appropriate type → `string | string[]`

File: src/typeahead/highlight.ts [hunk 3]
```diff
@@ -30,28 +30,30 @@ export class NgbHighlight implements OnChanges {
...
   ngOnChanges(changes: SimpleChanges) {
     const resultStr = toString(this.result);
-    const resultLC = resultStr.toLowerCase();
-    const termLC = toString(this.term).toLowerCase();
-    let currentIdx = 0;
+    if (!resultStr) {
+      this.parts = [this.result];
+      return;
+    }
+    let resultTerms: string[] = Array.isArray(this.term) ? this.term.map(x => toString(x)) : [toString(this.term)];
```
Comment (maxokorokov) [inline src/typeahead/highlight.ts]: Could you please also check the previous PR and my comment → https://github.com/ng-bootstrap/ng-bootstrap/pull/2750/files

I think combining the previous one with yours could make the code simpler ?

File: src/typeahead/highlight.ts (hunk 1 above)
Comment (IAfanasovMob) [reply src/typeahead/highlight.ts]: `any` probably is too much freedom. I see we have test coverage for string and number. Is it ok to change the type to `string | number`?

File: src/typeahead/highlight.ts (hunk 2 above)
Comment (IAfanasovMob) [reply src/typeahead/highlight.ts]: how about `string | string[] | number | number[]`? will ensure test coverage for all options

File: src/typeahead/highlight.ts (hunk 3 above)
Comment (IAfanasovMob) [reply src/typeahead/highlight.ts]: thanks @maxokorokov ! will check the previous attempt and get back with best wisdom from both prs.

**Comment by IAfanasovMob** [comment]
@maxokorokov I made a commit to address comments. want to squash it with an appropriate commit message after approval

**Comment by IAfanasov** [comment]
In vacation May 4-15. probably will find time for small fixes before. otherwise will be able to finish it after May 15

**Comment by IAfanasovMob** [comment]
@maxokorokov any more ideas on how to improve this pr? (:

**Comment by IAfanasov** [comment]
In the middle of the hight I realized number to string conversion is locale specific. Number '1234' could become '1 234' string. it leads to some confusion with highlighting '12'. I'm not ready to introduce such changes. Removed number from term and result input params.

**Comment by IAfanasovMob** [comment]
Thanks, @maxokorokov ! Do appreciate your work! (: 

> Thanks, @IAfanasov!
> 
> I'll merge it if you don't mind, but will simplify the implementation a bit as said in https://github.com/ng-bootstrap/ng-bootstrap/pull/3154/files#r279330567.

I try hard to grasp how 

```
let termsLC = Array.isArray(this.term) ? this.term : [this.term];
    termsLC = termsLC
      .map(term => regExpEscape(toString(term).toLowerCase()))
      .filter(term => term.length > 0);
// later

new RegExp(`(${termsLC.join('|')})`)
```
is simpler than 
```
    let resultTerms: string[] = Array.isArray(this.term) ? this.term.map(x => toString(x)) : [toString(this.term)];
    resultTerms = resultTerms.filter(x => x);
// later

    const regexStr = `(${resultTerms.map(x => regExpEscape(x)).join('|')})`;
```

At the proposed version we need to take care of making original text lower case and later restore it. Seems like a bit more complicated (: 
Also I didn't make `.map(x => regExpEscape(x)` right away after `let resultTerms: string[] = Array.isArray(this.term) ? this.term.map(x => toString(x)) : [toString(this.term)];` because `resultTerms` could become empty and `regExpEscape` will not be needed.

To speed things up - I totally agree to merge it if my name will be in the git history (:

**Comment by maxokorokov** [comment]
@IAfanasov sure, I will merge this one anyway! Thanks again.

I'm talking about simpler in terms of maintenance, not performance :)

This:
```ts
  ngOnChanges(changes: SimpleChanges) {
    const resultStr = toString(this.result);
    if (!resultStr) {
      this.parts = [resultStr];
      return;
    }
    let resultTerms: string[] = Array.isArray(this.term) ? this.term.map(x => toString(x)) : [toString(this.term)];

    resultTerms = resultTerms.filter(x => x);
    if (!resultTerms.length) {
      this.parts = [resultStr];
      return;
    }

    const regexStr = `(${resultTerms.map(x => regExpEscape(x)).join('|')})`;
    this.parts = resultStr.split(new RegExp(regexStr, 'gmi'));
  }
```

Vs this:
```ts
  ngOnChanges(changes: SimpleChanges) {
    const result = toString(this.result);
    let terms = Array.isArray(this.term) ? this.term : [this.term];

    terms = terms
      .map(term => regExpEscape(toString(term)))
      .filter(term => term.length);

    this.parts = !terms.length ? [result] : result.split(new RegExp(`(${terms.join('|')})`, 'gmi'));
  }
```

**Comment by IAfanasovMob** [comment]
Thanks for clarifications @maxokorokov ! (: I clearly see it's way simpler (: In my mind, the performance for the reusable library is more important than supportability. Probably it's just luck of open source contribution experience.

Happy to have my first proper open source contribution! (:

**Pull request merged**